### ISSUE TITLE
Crash on enchanting books and when books are generated in the world

### ISSUE THREAD

**Issue by JamesDClayton** [auoeke/limitless#34]
There is a complete crash when you try to enchant books and when you open a chest with an enchanted book inside.
Crash log is as follows
---- Minecraft Crash Report ----
// Why is it breaking :(

Time: 2/2/21 8:38 PM
Description: Watching Server

java.lang.Error: Watchdog
        at user11681.limitless.enchantment.EnchantmentUtil.getHighestSuitableLevel(EnchantmentUtil.java:26)
        at net.minecraft.class_1890.method_8229(class_1890.java:387)
        at net.minecraft.class_1890.method_8230(class_1890.java:337)
        at net.minecraft.class_1890.method_8233(class_1890.java:295)
        at net.minecraft.class_106.method_522(class_106.java:34)
        at net.minecraft.class_120.method_521(class_120.java:31)
        at net.minecraft.class_120.apply(class_120.java:20)
        at net.minecraft.class_117.method_514(class_117.java:14)
        at net.minecraft.class_117$$Lambda$8122/782207867.accept(Unknown Source)
        at net.minecraft.class_77.method_433(class_77.java:33)
        at net.minecraft.class_85$1.method_426(class_85.java:59)
        at net.minecraft.class_55.method_345(class_55.java:79)
        at net.minecraft.class_55.method_341(class_55.java:95)
        at net.minecraft.class_52.method_328(class_52.java:73)
        at net.minecraft.class_52.method_320(class_52.java:82)
        at net.minecraft.class_52.method_319(class_52.java:87)
        at net.minecraft.class_52.method_329(class_52.java:106)
        at net.minecraft.class_2621.method_11289(class_2621.java:81)
        at net.minecraft.class_2621.method_5438(class_2621.java:98)
        at net.minecraft.class_1264.method_5450(class_1264.java:26)
        at net.minecraft.class_1264.method_5451(class_1264.java:17)
        at net.minecraft.class_2281.method_9536(class_2281.java:190)
        at net.minecraft.class_4970$class_4971.method_26197(class_4970.java:870)
        at net.minecraft.class_2818.method_12010(class_2818.java:276)
        at net.minecraft.class_1937.method_30092(class_1937.java:231)
        at net.minecraft.class_1937.method_8652(class_1937.java:217)
        at net.minecraft.class_1937.method_8501(class_1937.java:321)
        at draylar.battletowers.entity.block.LadderDeployerBlockEntity.method_16896(LadderDeployerBlockEntity.java:49)
        at net.minecraft.class_1937.method_18471(class_1937.java:510)
        at net.minecraft.class_3218.method_18765(class_3218.java:429)
        at net.minecraft.server.MinecraftServer.method_3813(MinecraftServer.java:871)
        at net.minecraft.class_3176.method_3813(class_3176.java:312)
        at net.minecraft.server.MinecraftServer.method_3748(MinecraftServer.java:811)
        at net.minecraft.server.MinecraftServer.method_29741(MinecraftServer.java:670)
        at net.minecraft.server.MinecraftServer.method_29739(MinecraftServer.java:257)
        at net.minecraft.server.MinecraftServer$$Lambda$7168/234699379.run(Unknown Source)
        at java.lang.Thread.run(Thread.java:748)

if you need a list of mods i am using i will happily add them.

**Comment by user11681** [comment]
I believe that you missed a few lines of the error message above. It looks like generating enchantments takes too long; this is almost certainly a mod incompatibility. What version of limitless and what other mods do you use?

**Comment by JamesDClayton** [comment]
here is the full crash. 
I am using limitless 1.1.2
the mods are listed at the end of the report
[https://pastebin.com/emFegV0A](https://pastebin.com/emFegV0A)

**Comment by user11681** [comment]
I have released a [debug version](https://github.com/user11681/limitless/releases/tag/1.1.2-debug) of the mod, which should output some information that could help me locate the issue. Try triggering the crash with it. I need the entire server log.

**Comment by user11681** [comment]
I just noticed that I made a small mistake but it should still work.

**Comment by JamesDClayton** [comment]
The latest log with crash on debug
[https://pastebin.com/1n4VXG79]https://pastebin.com/1n4VXG79)

**Comment by user11681** [comment]
Hmm. It could be that the configured levels of some enchantments are too high; specifically, I suspect Channeling. You can reset the configuration by removing the configuration file `limitless.json5` or manually through the configuration screen through Mod Menu by resetting each enchantment entry in the per-enchantment configuration list.

**Comment by user11681** [comment]
I think that the problem is that the maximum level of an enchantment that does not normally scale (like Channeling, Flame and Infinity) is too high. The algorithm tries enchantment levels until the level with the highest affordable power is found, but if the power never changes, then the time taken depends on the maximum level for that enchantment, which can cause issues like this.

I will try to optimize this function when I should have more time.

**Comment by TelepathicGrunt** [comment]
A user on my mod page has reported a very similar worldgen crash caused by `_getOriginalMaxLevel() - min` being negative or 0.
https://www.curseforge.com/minecraft/mc-mods/repurposed-structures-fabric?comment=211

The line that is crashing for this user:
https://github.com/user11681/limitless/blob/77bb3ba2776d68123ff726f49aa801237aec325b/src/main/java/user11681/limitless/asm/mixin/enchantment/EnchantRandomlyLootFunctionMixin.java#L20

Curseforge comment:
```
Seems to crash if you create a world with a mineshaft near spawn


Full Crash Report:
https://pastebin.com/Q8As9wPB

Example seed [-1102463852006817912]
Default world settings.
With Traverse the only world gen mod.
Spawn in and head/turn to the left.

Every time on creation, world crashes with same error shortly after spawning.
If world is recreated without this mod, world does not crash.

Removal of mod does not fix crash on already generated world.
But will not cause crash if world is regenerated without mod.

Edit: Checked if it was due to the recent update, it is not
crash is present in 1.8.1 aswell.
```

**Comment by user11681** [comment]
I tried to reproduce this crash with Repurposed Structures, but I was unable to do so—perhaps as a result of unreleased changes—and I definitely am not going to test with all of these mods. I did make a change to `EnchantRandomlyLootFunctionMixin`, that might fix the issue, though. I also stopped depending on GrossFabricHacks. If version 1.2.0 does not fix the issue, then I am willing to look further into it so long as a minimal mod list is provided.

**Comment by BrekiTomasson** [comment]
Still having a similar issue in v1.2.0:

```
java.lang.NoSuchMethodError: net.minecraft.class_1887.getOriginalMaxLevel()I
	at Not Enough Crashes deobfuscated stack trace.(1.16.5+build.4)
	at net.minecraft.village.TradeOffers$EnchantBookFactory.create(TradeOffers:835)
	at net.minecraft.entity.passive.MerchantEntity.fillRecipesFromPool(MerchantEntity:261)
	at net.minecraft.entity.passive.VillagerEntity.fillRecipes(VillagerEntity:891)
	at net.minecraft.entity.passive.MerchantEntity.getOffers(MerchantEntity:115)
	at net.minecraft.entity.passive.VillagerEntity.interactMob(VillagerEntity:332)
	at net.minecraft.entity.mob.MobEntity.interact(MobEntity:1136)
	at net.minecraft.entity.player.PlayerEntity.interact(PlayerEntity:1049)
	at net.minecraft.client.network.ClientPlayerInteractionManager.interactEntity(ClientPlayerInteractionManager:382)
	at net.minecraft.client.MinecraftClient.doItemUse(MinecraftClient:1437)
	at net.minecraft.client.MinecraftClient.handleInputEvents(MinecraftClient:1709)
	at net.minecraft.client.MinecraftClient.tick(MinecraftClient:1534)
	at net.minecraft.client.MinecraftClient.render(MinecraftClient:1021)
	at net.minecraft.client.MinecraftClient.run(MinecraftClient:681)
	at net.minecraft.client.main.Main.main(Main:215)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.fabricmc.loader.game.MinecraftGameProvider.launch(MinecraftGameProvider.java:226)
	at net.fabricmc.loader.launch.knot.Knot.init(Knot.java:139)
	at net.fabricmc.loader.launch.knot.KnotClient.main(KnotClient.java:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196)
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231)
	at org.multimc.EntryPoint.listen(EntryPoint.java:143)
	at org.multimc.EntryPoint.main(EntryPoint.java:34)
```

The only other Enchanting-related mods that I've got installed are Charm, MC Dungeon Armors and MC Dungeon Weapons.

**Comment by user11681** [comment]
An oversight on my part. Should be fixed in 1.2.1.
